ddtrace/internal/injection.py

```diff
@@ -43,14 +44,35 @@ def _inject_hook(code, hook, lineno, arg):
     # >>> hook(arg)
     # Additionally, we must discard the return value (top of the stack) to
     # restore the stack to the state prior to the call.
-    code[i:i] = Bytecode(
-        [
-            Instr("LOAD_CONST", hook, lineno=lineno),
-            Instr("LOAD_CONST", arg, lineno=lineno),
-            Instr("CALL_FUNCTION", 1, lineno=lineno),
-            Instr("POP_TOP", lineno=lineno),
-        ]
-    )
+    if sys.version_info[:2] < (3, 11):
+        code[i:i] = Bytecode(
+            [
+                Instr("LOAD_CONST", hook, lineno=lineno),
+                Instr("LOAD_CONST", arg, lineno=lineno),
+                Instr("CALL_FUNCTION", 1, lineno=lineno),
+                Instr("POP_TOP", lineno=lineno),
+            ]
+        )
+    else:
+        code[i:i] = Bytecode(
+            [
+                Instr("PUSH_NULL", lineno=lineno),
+                Instr("LOAD_CONST", hook, lineno=lineno),
+                Instr("LOAD_CONST", arg, lineno=lineno),
+                Instr("PRECALL", 1, lineno=lineno),
+                Instr("CALL", 1, lineno=lineno),
+                Instr("POP_TOP", lineno=lineno),
+            ]
+        )
+
+
+_INJECT_HOOK_OPCODES = (
+    ["LOAD_CONST", "LOAD_CONST", "CALL_FUNCTION", "POP_TOP"]
+    if sys.version_info < (3, 11)
+    else ["PUSH_NULL", "LOAD_CONST", "LOAD_CONST", "PRECALL", "CALL", "POP_TOP"]
+)
+_INJECT_HOOK_OPCODE_POS = 0 if sys.version_info < (3, 11) else 1
+_INJECT_ARG_OPCODE_POS = 1 if sys.version_info < (3, 11) else 2
```

Comment:
This is a really small thing. You could make this code slightly more future-proof by setting these based on the `_INJECT_HOOK_OPCODES.index()` of the particular instruction string these constants are supposed to indicate.

Reply:
That's a nice idea, but I think there's no way of telling how the Python bytecode will evolve 🙁